library/init/meta/tactic.lean

```diff
@@ -997,6 +997,20 @@ match _root_.try_for max (tac s) with
 | none   := mk_exception "try_for tactic failed, timeout" none s
 end
 
+meta def updateex_env (f : environment → exceptional environment) : tactic unit :=
+do env ← get_env,
+   env ← returnex $ f env,
+   set_env env
+
+meta def add_decl (d : declaration) : tactic unit :=
+updateex_env $ λe, do e ← e.add d, e.compile d tt
+
```

Comment:
It felt cleaner to provide more fine grained operations to the Lean side. I will revert `add_decl`, to be implemented in C++ again.